kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h

```diff
@@ -958,7 +958,7 @@ class ResidualBasedBlockBuilderAndSolver
     ///@name Protected Operations
     ///@{
 
-    void ConstructMasterSlaveConstraintsStructure(ModelPart& rModelPart)
+    virtual void ConstructMasterSlaveConstraintsStructure(ModelPart& rModelPart)

```

Comment:
Please, move this to the base builder and solver, otherwise other B&S not derived from the block cannot benefit from this (additionally cannot be called if not stated like this without casting the class). In my opinion should be public too, because it is a methos that can be interesting to call from outside. Maybe your not constrained system of equations keeps the same distribution, but you constraints may be non-linear and you want to update them. Of course for this case the aspect of the resulting matrix must be the same in ordeer to be called only from the constraint side, but any way I think that for consistency and better developments in the future is important

Reply:
I do agree about moving to the base b&s .. ( By the way there are no constraint related methods in the base B&S  and also other methods like `ConstructMatrixStructure` are also not there)
But I will suggest we do it in a separate PR .. this is because if we decide to add methods to the base builder and solver , it will include many more functions and not just the ones related to constraints. 

I am not sure if I understood what you mean the nonlinear constraint changes the distribution. 
Once the constraint is made , its slave and masters are fixed. A non-linear constraint only changes the relation between them.

Reply:
That's why I asked for an independent PR

Reply:
Any case, if non-linear means that can be activated inactivated, this chnages the structure. See contact

Reply:
Well its an independent PR now .. but still I think we should change those methods to base in a different PR. 
And the changes here are still valid in that case too.

Reply:
this is not true. The point is that when you construct the structure you do not check for active/inactive, so the graph contains the nonzeros and doesn't need to be modified.

Reply:
Do a test @RiccardoRossi, take a problem with MPC which does are not included on the element/conditions, deactivate in a step all the MPC, the problem will crash, because when assembling it will not find those unused dofs